[WIP] patcher concepts: what is a patch (#2180)
gruntwork-io/docs: docs/2.0/docs/patcher/concepts/patches.md

It is best practice to reference a specific version of an OpenTofu/Terraform module. 
Over time, module authors release new versions of the module, and the code that consumes those modules slowly gets out of date. 
In some cases, the latest update of the underlying modules requires a breaking change to the consuming code, meaning the version can't just be bumped. 
The consuming code needs to be edited to make use of the new version. 
In this instance, using a patch with Patcher comes in handy.

Patches can be consumed with either a "push" strategy, when Patcher proactively opens a pull request with the latest update, or a "pull" strategy, when a repo is manually scanned to look at the current state of your infrastructure using the Patcher CLI tool.

Regardless of methodology, the concept remains the same. 
Patcher will suggest changes to your codebase in order to keep your infrastructure up to date, however you see fit. 

Here is an example of what your PR will look like, after being generated by Patcher:

> ** insert image of PR screenshot **

Reviewer: TODO @ceschae

### Update Push Strategy

Let's take a look at the experience of the "push" strategy available through our [GitHub action](https://github.com/gruntwork-io/patcher-action). 
Using the GitHub action, PRs will be opened against your codebase on a cadence specified by you, against environments, versions, etc. specified by you. 
The intention with this GitHub action is to leave the repo owner in full control of your upgrade cadence. 

> <details>

Reviewer: I worry this will go stale, and is also too much unexplained detail for a concept page.  Maybe better to just remove it and link instead to the promotion workflows guides?

Author: just to confirm which noun we're talking about - remove "it" being the full example?

>   <summary>The example configuration is long. Click to expand!</summary>

Reviewer: This is a bit too casual IMO for our docs and doesn't match the style used elsewhere.  How about just something very direct "GitHub Actions Workflow Example"

          echo '${{ needs.patcher-report.outputs.spec }}' > patcher-spec.json

      - uses: gruntwork-io/patcher-action@main
        with:
          patcher_command: update
          github_token: ${{ secrets.PIPELINES_READ_TOKEN }}
          working_dir: ${{ env.ENV_FOLDER_NAME }}
          dependency: ${{ matrix.dependency.ID }}
          spec_file: patcher-spec.json
          pull_request_title: "[Patcher] Update ${{ matrix.dependency.ID }}"
          pull_request_branch: "${{ env.PR_BRANCH_PREFIX }}${{ matrix.dependency.ID }}"
  ```
</details>
<br/>

Also, check out our guide on [promotion workflows](../guides/promotion-workflows), so that updates can proceed from `dev` to `stage` to `prod` (or any other environment sequence) to mitigate risks around upgrades. 

### Update Pull Strategy

Let's take a look at the experience of "pulling" available updates by using the Patcher CLI tool. 
The first step is to run `patcher update` within the repo in which updates are desired. 
When `patcher update` is run, the default mode is to click through the updates **interactively**. 
In this mode, available updates are found, and the details of those updates are presented to you:

> ** insert image here **

Reviewer: @ceschae TODO

You can choose to run in `--non-interactive` mode, which will modify the codebase and present results about what the program did at the end. 

> Interactively nor not, by default, a PR will _not_ be opened with the changes.

Reviewer: ```suggestion
By default a PR will _not_ be opened with the changes.
```

However, the changes should be visible within the version control system. At that point, you may make a PR or apply the changes using your IaC system.

### Examples Running `patcher update`

Here's the easiest way to run this command:

```bash
$ cd <repo-root-directory>

> # give me ALL the patches for ALL the things

Reviewer: Again a bit too casual.
```suggestion
# Show what patches are available for everything in the current directory and all it's children
```

$ patcher update ./
```

If more fine-grain controls are desired, the following example has those:

```bash
# run 'update' non-interactively, only up to the next safe version, and publish a PR with the changes
$ patcher update --update-strategy next-safe --non-interactive --publish --pr-branch grunty/update-via-patcher --pr-title "[Patcher] Update All Dependencies to Next Safe"
```

More details on the available options included in `patcher update` can be found in the [reference section](/2.0/reference/patcher/index.md#update). 

## For Module Authors

Module authors periodically need to introduce breaking changes in their modules, causing a downstream, potentially painful, experience for module consumers. 
With patches, module authors include a patch YAML file that automatically updates consuming code to incorporate the breaking changes associated with the updated module code.

> Doing so allows module consumers to use patches to enable their modules consumers to automatically update consuming code to adopt breaking change.

Reviewer: ```suggestion
Doing so allows module consumers to use patches to enable their modules consumers to automatically update consuming code to adopt breaking changes.
```

In a Patcher ecosystem, the resolution to such a change is written once, in a patch, and distributed to all consumers. 
Although your release will succeed with or without a patch, downstream consumers of your breaking change will praise you thoroughly for your advance work. 

In theory, you may write whatever command execution steps you want to perform patch steps. 
For example, there are many cases where validating tool versions are required, or using `sed` to find and replace certain values. 

> However, we _strongly_ recommend using `terrapatch`, a Gruntwork tool that surgically updates Terraform/OpenTofu HCL files.  

Reviewer: ```suggestion
However, we _strongly_ recommend using [terrapatch](https://github.com/gruntwork-io/terrapatch), a Gruntwork tool that surgically updates Terraform/OpenTofu HCL files.  
```

# What is a Patcher Patch?

A Patcher Patch is a set of instructions that Patcher can carry out to resolve a breaking change. 
This instruction sheet is delivered by means of a `yaml` file in a specific format:

```yaml
name: "<name-of-patch>"
description: "<description for patch>"
author: <your-name-here> <<your-email-address-here>>

# Optional dependencies. Terrapatch is a typical one
dependencies:
  - name: terrapatch
    version: "0.1.0"

# Steps necessary to resolve breaking change
steps:
  - name: "<name-of-step>"
    run: <command-to-run>
  - name: "<name-of-second-step>"
    run: <second-command-to-run>
  # etc

> ```

Author: ope - no, i don't think that ever got written down in Linear, my b. here it is now: https://linear.app/gruntwork/issue/PAT-39/include-affected-modules-with-patchyaml-file-in-addition-to-configyaml

## As a Consumer of Patches...

All that's required to ingest patches for breaking IaC changes is to run a `patcher update` (with whatever flags you deem appropriate). 

> This command can be run locally, in which case, the steps outlined in the patch(es) would be executed within your codebase, and it would be up to you to apply IoC changes, or open them in a PR against your codebase, whichever is appropriate for your use case. 

Reviewer 2: This sounds confusing as written. Can you more clearly separate the specific things that are happening?

For example:
- `patcher update` executes the steps in the patch
- This changes the code in your local environment, but does not commit it or do any kind of `terragrunt apply` or other operation against it.
- It's up to you to apply the code and open a PR that submits it for review.

> Additionally, Gruntwork offers a GitHub Action to automate this process for you.

Reviewer 2: Use the "push" intro for this. Consider adding a header for pull and one for push.

Using the GitHub action, PRs will be opened against your codebase on a cadence specified by you, against environments, versions, etc. specified by you. 

> The intention with this GitHub action is to leave you, the repo owner, in full control of your upgrade cadence. 

Reviewer 2: Include PR screenshot?

Regardless of methodology, the concept remains the same. 
Patcher will suggest changes to your codebase in order to keep your infrastructure up to date, however you see fit. 

## As an Author of Patches...

> As an author of a patch, this means that you are publishing a breaking change with a release of your module. 

Reviewer 2: Switch to more active language. Instead of "this means that you are...," consider language like "Module authors periodically need to introduce breaking changes in their modules, causing a downstream painful experience for module consumers. With patches, module authors write not only the updated module code, but also include a patch YAML file that automatically updates consuming code to incorporate the breaking changes.

use patches to enable their modules consumers to automatically update consuming code to adopt breaking change.

Author: i like your verbage! incorporated 👍 

as a general note, you'll likely notice i abstain from assertive/aggressive language. if i swung too far away from the docs site's conventions, please do feel free to call me out, always happy to fix it!
